### PR TITLE
node_exporter 0.17.0

### PR DESCRIPTION
**Pull request by Josh-Tilles**
In order to adapt to the changes introduced in version 0.17.0,
I thought it best to embrace the build tooling included in the project
itself—namely, Makefiles and [the `promu` utility]. In addition,
the formula now runs the project’s tests during the build.

[the `promu` utility]: https://github.com/prometheus/promu

- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by MerelyAPseudonym** [comment]
A few other things I wanted to call out:
1. Using the `gnu-sed` formula reduced the build time on my system from ~35 seconds to ~10. I guessed the extra build-time dependency wasn’t worth it, but it’s up to you.
2. I also guessed that you’d prefer to manage `promu` as a resource rather than allow the Makefile to download it—but please feel free to let me know I guessed incorrectly 🙂
3. Caveat: I only dabble in the Go ecosystem. As such, I don’t know enough to be confident the solution I came up with is the best.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

File: Formula/node_exporter.rb [hunk 1]
```diff
@@ -14,13 +14,30 @@ class NodeExporter < Formula
 
   depends_on "go" => :build
 
+  resource "promu" do
+    # check and update upon next release
+    # (matches <https://github.com/prometheus/node_exporter/blob/v0.17.0/Makefile.common#L68-L69>)
+    url "https://github.com/prometheus/promu/releases/download/v0.2.0/promu-0.2.0.darwin-amd64.tar.gz"
```
Comment (fxcoudert) [inline Formula/node_exporter.rb]: Let's not download binaries **at all**, please.

File: Formula/node_exporter.rb [hunk 2]
```diff
@@ -14,13 +14,30 @@ class NodeExporter < Formula
...
+    sha256 "6372caea7d8e7c0f208f6c040e2bfb462c72c9f7b2f6c13b7992d7caa5fbcf75"
+  end
+
+  # Fix checksum mismatch of `soundcloud/go-runit@v0.0.0-20150630195641-06ad41a06c4a`.
+  patch do
+    url "https://github.com/prometheus/node_exporter/commit/97dab59e18c575d1384a9afd54e9f30473f34322.patch?full_index=1"
```
Comment (fxcoudert) [inline Formula/node_exporter.rb]: Why is that necessary? Please confirm the issue with upstream, to make sure it's also fixed on their end.

**Review by fxcoudert — CHANGES_REQUESTED** [review]

File: Formula/node_exporter.rb (hunk 1 above)
Comment (MerelyAPseudonym) [reply Formula/node_exporter.rb]: Do you mean that the formula should build the `promu` utility from source, or did you have something else in mind? Regardless, I'm happy to oblige!

File: Formula/node_exporter.rb (hunk 2 above)
Comment (MerelyAPseudonym) [reply Formula/node_exporter.rb]: [According to the patch author](https://github.com/prometheus/node_exporter/issues/1199):
> the golang checksum code changed between 1.11.3 and 1.11.4 to correctly handle symlinks, and thus broke `go.sum` for a bunch of projects.

And upstream is definitely aware—the patch is from a commit already present in the project’s `master` branch. (And so the patch can be removed with the next release of `node_exporter`.)

Is that sufficient? Would you prefer for the explanatory comment to have more detail?

**Comment by MerelyAPseudonym** [comment]
@fxcoudert: this is a polite ping to make sure that my replies and questions hadn't flown under your radar. (I.e., no sweat if you're just busy.)

**Comment by SMillerDev** [comment]
> 1. Using the `gnu-sed` formula reduced the build time on my system from ~35 seconds to ~10. I guessed the extra build-time dependency wasn’t worth it, but it’s up to you.

Most users won't experience that as they'll use the bottled version while it is a part that can break.

> 2. I also guessed that you’d prefer to manage `promu` as a resource rather than allow the Makefile to download it—but please feel free to let me know if I guessed incorrectly 🙂

Yes, downloads should be done as resources 

> 3. Caveat: I only dabble in the Go ecosystem. As such, I don’t know enough to be confident the solution I came up with is the best.

Me neither 😅

File: Formula/node_exporter.rb (hunk 1 above)
Comment (fxcoudert) [reply Formula/node_exporter.rb]: Yes, let's build from source.

File: Formula/node_exporter.rb (hunk 2 above)
Comment (fxcoudert) [reply Formula/node_exporter.rb]: OK, please add link to the issue in comment

**Review by fxcoudert — CHANGES_REQUESTED** [review]

**Comment by chenrui333** [comment]
I think this can be closed in favor of a merged PR, https://github.com/Homebrew/homebrew-core/pull/40024

**Comment by chenrui333** [comment]
I also submit a fresh PR in here, https://github.com/Homebrew/homebrew-core/pull/40694
